library/core/src/array/iter.rs

```diff
@@ -123,6 +123,22 @@ impl<T, const N: usize> Iterator for IntoIter<T, N> {
         (len, Some(len))
     }
 
+    #[inline]
+    fn fold<Acc, Fold>(mut self, init: Acc, mut fold: Fold) -> Acc
+    where
+        Fold: FnMut(Acc, Self::Item) -> Acc,
+    {
+        let data = &mut self.data;
+        (&mut self.alive)
+            .try_fold::<_, _, Result<_, !>>(init, |acc, idx| {
+                // SAFETY: idx is obtained by folding over the `alive` range, which implies the
+                // value is currently considered alive but as the range is being consumed each value
+                // we read here will only be read once and then considered dead.
+                Ok(fold(acc, unsafe { data.get_unchecked(idx).assume_init_read() }))
+            })
+            .unwrap()
```

Comment:
can we call `fold` here instead of `try_fold`? 

```suggestion
        self.alive.fold(init, |acc, idx| {
            // SAFETY: idx is obtained by folding over the `alive` range, which implies the
            // value is currently considered alive but as the range is being consumed each value
            // we read here will only be read once and then considered dead.
            fold(acc, unsafe { data.get_unchecked(idx).assume_init_read() })
        })
```

Reply:
I tried that, `array::IntoIter` has a Drop impl, so `alive` can't be move out, but that would be required to call `fold(self)`, that's why I used `try_fold instead`.

Reply:
@the8472 oops right.

`(&mut self.alive).fold(init, ...)` should work though.

Reply:
Yeah but that would go through `impl Iterator for &mut I` which is less optimized.

Reply:
what about `std::mem::take(&mut self.alive).fold(...)`?

Reply:
Alive is needed by the drop impl, so if the fold paniced then there would be a problem. Maybe it could be put back in place before the unwind tries to drop it, but that would require a drop guard which would make things more complicated instead of simpler.

Reply:
`(&mut x).fold` handles the panic just fine, with the same behavior as `x.try_fold`.

```rust
use std::ops::Range;

struct A {
    a: Range<usize>,
}

impl Drop for A {
    fn drop(&mut self) {
        dbg!(&self.a); // should print 5..7 instead of 0..7
    }
}

fn main() {
    let mut a = A { a: 0..7 };

    (&mut a.a).fold((), |(), i| {
        dbg!(i);
        if i == 4 {
            panic!();
        }
    });
}
```

Reply:
> (&mut x).fold handles the panic just fine

My previous comment was replying to the `mem::take` suggestion which was edited away after I replied.

> self.alive is a std::ops::Range<usize> and AFAIK there is no special-cased implementation of fold or try_fold for Range<usize> nor &mut Range<usize>.

Ah, it was  `RangeInclusive` that had one but not `Range`. So yeah that might work. I'll give it a try.

Reply:
even if we do the `mem::take` it will just turn `self.alive` to `0..0` and then leaks everything which is safe 🙃  (compared with `self.alive.clone().fold(...)` which will cause double-free).

Reply:
I have now tried `(&mut self.alive).fold` instead of `try_fold`, it undoes all perfomance gains. I guess somehow the indirection through `&mut` inhibits optimizations.

Reply:
Which is due to this not being `#[inline]`

https://github.com/rust-lang/rust/blob/71a6c7c80398143a67e27b01412f4b2ec12bde8b/library/core/src/iter/traits/iterator.rs#L3474-L3478

Reply:
heh. 

wdyt should we just add the `#[inline]` or leave a FIXME comment explaining the performance regression if we use `fold` instead of `try_fold`? either way is fine for me.

Reply:
I'll add a FIXME, changing inlining on such central methods can have mixed impact on compile time even if runtime performance is better, so that should be done on a separate PR.